File: framework/ui/src/io/intino/alexandria/ui/displays/components/SelectorTabs.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.SelectorTabsOptionVisibility;
import io.intino.alexandria.ui.displays.components.selector.SelectorOption;
import io.intino.alexandria.ui.displays.notifiers.SelectorTabsNotifier;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SelectorTabs<DN extends SelectorTabsNotifier, B extends Box> extends AbstractSelectorTabs<DN, B> {
	private int selected = -1;

    public SelectorTabs(B box) {
        super(box);
    }

	@Override
	public void didMount() {
		super.didMount();
		selection(selected);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<String> selection() {
		if (selected == -1) return Collections.emptyList();
		return Collections.singletonList(nameOf(selected));
	}

	public void selection(String option) {
		int position = position(option);
		if (position == -1) return;
		selection(position);
	}

	public void selection(int option) {
		notifier.refreshSelected(option);
		if (this.selected == option) return;
		this.selected = option;
	}

	public void select(String option) {
    	selection(option);
		notifySelection();
	}

	public void selectByName(String option) {
    	select(option);
	}

	public void select(int option) {
    	selection(option);
		notifySelection();
	}

	public void showOption(String option) {
		hiddenOptions.remove(option);
		notifier.refreshOptionsVisibility(optionsVisibility());
	}

	public void hideOption(String option) {
		hiddenOptions.add(option);
		notifier.refreshOptionsVisibility(optionsVisibility());
	}

	public void show(String option) {
		SelectorOption component = findOption(position(option));
		component.visible(true);
	}

	public void hide(String option) {
		SelectorOption component = findOption(position(option));
		component.visible(false);
	}

	@Override
	public void reset() {
		selected = -1;
		notifier.refreshSelected(null);
	}

	private List<SelectorTabsOptionVisibility> optionsVisibility() {
		List<SelectorTabsOptionVisibility> result = new ArrayList<>();
		List<SelectorOption> options = options();
		for (int i=0; i<options.size(); i++) result.add(new SelectorTabsOptionVisibility().index(i).visible(!hiddenOptions.contains(options.get(i).name())));
		return result;
	}

}

File: framework/ui-elements/src/displays/components/SelectorTabs.js
import React from "react";
import {withStyles} from 'alexandria-ui-elements/src/util/muiStylesCompat';
import {Tab, Tabs} from '@mui/material';
import AbstractSelectorTabs from "../../../gen/displays/components/AbstractSelectorTabs";
import SelectorTabsNotifier from "../../../gen/displays/notifiers/SelectorTabsNotifier";
import SelectorTabsRequester from "../../../gen/displays/requesters/SelectorTabsRequester";
import Divider from './Divider';
import DisplayFactory from 'alexandria-ui-elements/src/displays/DisplayFactory';
import {withSnackbar} from "alexandria-ui-elements/src/util/notistackCompat";

const styles = theme => ({
	root: theme.palette.mode === "dark" ? {
		minHeight: "36px",
		padding: "0 8px",
	} : {},
	indicator: {
		height: "2px",
		borderRadius: "999px",
		backgroundColor: theme.palette.primary.main,
	},
	tabRoot: theme.palette.mode === "dark" ? {
		minHeight: "36px",
		paddingTop: 0,
		paddingBottom: 0,
		color: "rgba(226,232,240,0.72)",
		"&.Mui-selected": {
			color: "#90caf9",
		},
		"&.Mui-disabled": {
			opacity: 0.4,
		}
	} : {}
});

class SelectorTabs extends AbstractSelectorTabs {

	constructor(props) {
		super(props);
		this.notifier = new SelectorTabsNotifier(this);
		this.requester = new SelectorTabsRequester(this);
		this.state = {
		    ...this.state,
            selected: -1,
            optionsVisibility: []
        };
	};

	render() {
	    if (!this.state.visible) return (<React.Fragment/>);
	    const { classes } = this.props;
	    const selected = this.state.selected !== -1 ? this.state.selected : 0;
	    const children = this._visibleOptions();
	    const scrollButtons = this.props.scrollButtons != undefined ? this.props.scrollButtons.toLowerCase() : "off";
	    const variant = scrollButtons !== "off" ? "scrollable" : undefined;
        return (
            <Tabs value={selected} variant="fullWidth" variant={variant} scrollButtons={scrollButtons}
                  onChange={this.handleChange.bind(this)} color={this.props.color} style={this.style()}
                  className={classes.root} classes={{ indicator: classes.indicator }}>
                {React.Children.map(children, (child, i) => { return this.renderTab(child, i); })}
            </Tabs>
        );
    }

    renderTab = (tab, i) => {
	    const className = tab.props.className;
        if (className != null && className.indexOf("divider") !== -1) return (<Divider/>);
        return (<Tab label={tab} index={i} style={this.styleOf(tab)} className={this.props.classes.tabRoot}/>);
    };

    refreshSelected = (tab) => {
        let realTab = tab - this._countInvisible(tab);
        this.setState({ selected: realTab });
	};

    refreshOptionsVisibility = (optionsVisibility) => {
        const visibility = [];
        for (var i=0; i<optionsVisibility.length; i++) visibility[optionsVisibility[i].index] = optionsVisibility[i].visible;
        this.setState({ optionsVisibility: visibility });
	};

	_countInvisible = (pos) => {
	    const optionsVisibility = this.state.optionsVisibility.length > 0 ? this.state.optionsVisibility : this._defaultOptionsVisibility();
	    let result = 0;
	    for (var i=0; i<optionsVisibility.length; i++) {
	        if (i <= pos && optionsVisibility[i] === false) result++;
	    }
	    return result;
	};

	_defaultOptionsVisibility = () => {
	    const children = this.children();
	    const result = [];
	    for (var i=0; i<children.length; i++) {
	        const name = children[i].props.name;
	        result[i] = this.props.hiddenOptions == null || this.props.hiddenOptions.indexOf(name) == -1;
	    }
	    return result;
	};

	_visibleOptions = () => {
	    const children = this.children();
	    const result = [];
	    for (var i=0; i<children.length; i++) {
	        if (!this._isVisible(children[i], i)) continue;
	        result.push(children[i]);
	    }
	    return result;
	}

	_isVisible = (tab, pos) => {
	    const optionsVisibility = this.state.optionsVisibility;
	    return optionsVisibility.length > 0 ? optionsVisibility[pos] : (tab != null && tab.props != null && (tab.props.visible == null || tab.props.visible !== false));
	}

	handleChange = (e, value) => {
	    const children = this._visibleOptions();
	    let selected = children[value] != null ? children[value].props.name : null;
	    if (selected == null && !(children instanceof Array) && children.props != null) selected = children.props.name;
	    if (selected != null) this.requester.selectByName(selected);
	    else this.requester.select(value);
    };
}

export default withStyles(styles, { withTheme: true })(withSnackbar(SelectorTabs));
DisplayFactory.register("SelectorTabs", withStyles(styles, { withTheme: true })(withSnackbar(SelectorTabs)));
